### PR TITLE
Implementation of Parallel Merkle Tree

### PR DESCRIPTION
**Pull request by intx4**
## Description

This PR implements a Parallel Version of the Merkle Tree, completely transparent to the user, i.e. it defines two new versions of `MerkleTree::new(...)` and `MerkleTree::new_with_leaf_digest(...)` which are compiled (in place of the legacy ones) when the "parallel" feature is enabled. The implementation can be found in `src/merkle_tree/mod.rs`. We have also included some benchmarks in `benches/merkle_tree.rs`.
Using our implementation, for a Merkle Tree with `2^20` leaves, using SHA256, we obtain a speedup of 80% over legacy on a M1 MacBook.

closes: #122 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by WizardOfMenlo** [comment]
There is currently a change of semantics between the parallel and non-parallel version. Namely, if one of the hash functions (the leaf hash or the two to one hash) return `Err`, the serial version would return it correctly to the caller, while the parallel version would panic. This can be fixed easily enough as long as the error type bound is changed to `dyn Error + Send`. 
1. Is it acceptable to change API to add the `Send` bound for the parallel bound? Or even, can we have `ark_std::Error` to require `Send` by default?
2. When is it desirable for the `LeafHash, TwoToOneHash` to be allowed to fail? In most use-cases I can think of the hashing should always succeed. I would be in favour to refactor those traits to remove the error path. The conversion traits I can see more usecases, but also wouldn't mind changing the API there as well.

@mmagician @Pratyush what are your thoughts?

File: src/merkle_tree/mod.rs [hunk 1]
```diff
@@ -234,26 +239,145 @@ impl<P: Config> MerkleTree<P> {
...
+        let leaves_digest: Result<Vec<_>, _> = leaves
+            .into_iter()
+            .map(|leaf| P::LeafHash::evaluate(leaf_hash_param, leaf))
+            .collect();
+
+        Self::new_with_leaf_digest(leaf_hash_param, two_to_one_hash_param, leaves_digest?)
+    }
+
+    /// Returns a new merkle tree. `leaves.len()` should be power of two.
+    #[cfg(feature = "parallel")]
```
Comment (mmagician) [inline src/merkle_tree/mod.rs:259]: Maybe you can avoid having a separate block by using `cfg_into_iter`: https://github.com/arkworks-rs/std/blob/master/src/lib.rs#L136
See example of usage here: https://github.com/arkworks-rs/algebra/blob/ab13aa09ae3c11cde0224028dee7b878bbcf9246/ec/src/models/mnt6/mod.rs#L56

File: src/merkle_tree/tests/mod.rs [hunk 2]
```diff
@@ -47,35 +50,78 @@ mod bytes_mt_tests {
...
-                .verify(&leaf_crh_params, &two_to_one_params, &root, leaf.as_slice())
-                .unwrap());
+
+        #[cfg(feature = "parallel")]
+        {
+            // Parallel
+            let mut tree = JubJubMerkleTree::new(
+                &leaf_crh_params.clone(),
+                &two_to_one_params.clone(),
+                leaves.par_iter().map(|x| x.as_slice()),
```
Comment (mmagician) [inline src/merkle_tree/tests/mod.rs:97]: Same here I think

File: src/merkle_tree/mod.rs [hunk 3]
```diff
@@ -234,26 +239,145 @@ impl<P: Config> MerkleTree<P> {
...
+                //to get the index outside the slice and in the level-ordered list of nodes
+
+                let current_index = i + start_index;
+                let left_leaf_index = left_child(current_index) - upper_bound;
+                let right_leaf_index = right_child(current_index) - upper_bound;
+                // compute hash
+                *n = P::TwoToOneHash::evaluate(
+                    two_to_one_hash_param,
+                    P::LeafInnerDigestConverter::convert(leaves_digest[left_leaf_index].clone())
+                        .unwrap(),
```
Comment (mmagician) [inline src/merkle_tree/mod.rs:330]: can't we propagate the errors here?

File: src/merkle_tree/mod.rs [hunk 4]
```diff
@@ -234,26 +239,145 @@ impl<P: Config> MerkleTree<P> {
...
+                    P::LeafInnerDigestConverter::convert(leaves_digest[right_leaf_index].clone())
+                        .unwrap(),
+                )
+                .unwrap()
```
Comment (mmagician) [inline src/merkle_tree/mod.rs:334]: and here

**Comment by mmagician** [comment]
Rather than duplicating the entire method for the `parallel` feature, I would instead try to re-use as much code as possible and only but blocks of code under `#[cfg(feature = "parallel")]` and `#[cfg(not(feature = "parallel"))]`

**Comment by intx4** [comment]
@mmagician thanks for your feedback, I will address them and get back to you

File: src/merkle_tree/mod.rs (hunk 4 above)
Comment (intx4) [reply src/merkle_tree/mod.rs:334]: Unfortunately we cannot without introducing some changes to the API, refer to @WizardOfMenlo comment above https://github.com/arkworks-rs/crypto-primitives/pull/125#issuecomment-1867841425

**Comment by intx4** [comment]
> @mmagician thanks for your feedback, I will address them and get back to you

@mmagician I should have addressed your comments, feel free to get back with feedback :)

File: benches/merkle_tree.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                //}
+                //#[cfg(feature = "parallel")]
+                //{
+                //    _ = Sha256MerkleTree::new(
+                //        &leaf_crh_params.clone(),
+                //        &two_to_one_params.clone(),
+                //        leaves.par_iter().map(|x| x.as_slice()),
+                //    )
+                //    .unwrap();
+                //}
```
Comment (Pratyush) [inline benches/merkle_tree.rs:70]:
```suggestion
```

**Comment by Pratyush** [comment]
I think it's fine to change the `crate::Error` from `Box<dyn ark::std::Error>` to `Box<dyn ark::std::Error + Send>`. That should solve the issues around unwrapping.

**Comment by Pratyush** [comment]
btw @intx4 could you allow edits to this PR by repo maintainers? I have a few small nits that I'd like to push directly to your branch. Thanks!

**Review by Pratyush — APPROVED** [review]

**Pull request merged**